### PR TITLE
ISSUE-133 Add wish list imap commands

### PR DESCRIPTION
**Pull request by quantranhong1999**
Wish list to be feature complete on IMAP:

- [x]  Subscribe
- [x]  Unsubscribe
- [x]  Create
- [x]  Delete
- [x]  Rename
- [x]  Examine
- [x] Move
- [x]  Copy
- [x]  Uid expunge
- [x]  Uid Store
- [x]  Uid Move
- [x]  Uid copy
- [x]  Uid search
- [x]  Get quota
- [x]  Set quota
- [x]  COMPRESS

### CODE REVIEW

File: src/main/scala-2.12/com/linagora/gatling/imap/scenario/ImapGetQuotaScenario.scala [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+import io.gatling.core.structure.ScenarioBuilder
+
+object ImapGetQuotaScenario {
+  def apply(feeder: FeederBuilder): ScenarioBuilder =
+    scenario("Imap")
+      .feed(feeder)
+      .exec(imap("Connect").connect()).exitHereIfFailed
+      .exec(imap("login").login("${username}", "${password}").check(ok))
+      .exec(imap("select").select("INBOX").check(ok))
+      .exec(imap("getQuota").getQuota("\"" + "\"").check(ok))
```
Comment (Arsnael) [inline src/main/scala-2.12/com/linagora/gatling/imap/scenario/ImapGetQuotaScenario.scala:15]:
```suggestion
      .exec(imap("getQuota").getQuota("\"\"").check(ok))
```
?

File: src/main/scala-2.12/com/linagora/gatling/imap/scenario/ImapGetQuotaScenario.scala [hunk 2]
```diff
@@ -12,5 +12,6 @@ object ImapGetQuotaScenario {
       .exec(imap("Connect").connect()).exitHereIfFailed
       .exec(imap("login").login("${username}", "${password}").check(ok))
       .exec(imap("select").select("INBOX").check(ok))
+      .exec(imap("setQuota").setQuota("INBOX (STORAGE 512)").check(no))
       .exec(imap("getQuota").getQuota("INBOX").check(bad))
```
Comment (quantranhong1999) [inline src/main/scala-2.12/com/linagora/gatling/imap/scenario/ImapGetQuotaScenario.scala:16]: Could this no and bad be acceptable? (I do not fully understand set/get quota though).
James response:
```
a3 NO SETQUOTA You need the Full admin rights right to perform command SETQUOTA on mailbox Can not perform SETQUOTA commands.

a4 BAD GETQUOTA failed. No such mailbox.
```

Cyrus response:
```
a3 NO Permission denied

a4 NO Quota root does not exist
```

This does not come from the 2 commands themselves. Any suggestion about the scenario is welcome :')

Comment (Arsnael) [reply src/main/scala-2.12/com/linagora/gatling/imap/scenario/ImapGetQuotaScenario.scala:16]: I don't know about Cyrus, but in James I would guess you can try to add a usersrepository.xml in james conf, with the `administratorId` declared to the user you would use in your integration test. Then that user should have admin rights to set up that quota? My guess, but not 100% sure :)

File: src/main/scala-2.12/com/linagora/gatling/imap/action/GetQuotaAction.scala [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+object GetQuotaAction {
+  def props(imapContext: ImapActionContext, requestname: String, checks: Seq[ImapCheck], mailbox: Expression[String]) =
+    Props(new GetQuotaAction(imapContext, requestname, checks, mailbox))
+}
+
+class GetQuotaAction(val imapContext: ImapActionContext, val requestName: String, override val checks: Seq[ImapCheck], mailbox: Expression[String]) extends ValidatedActionActor with ImapActionActor {
+
+  override protected def executeOrFail(session: Session): Validation[_] = {
+    for {
+      mailbox <- mailbox(session)
```
Comment (chibenwa) [inline src/main/scala-2.12/com/linagora/gatling/imap/action/GetQuotaAction.scala:21]: Parameter of get/set quota is a quotaRoot and not a mailbox...

File: src/main/scala-2.12/com/linagora/gatling/imap/protocol/command/SubscribeHandler.scala [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+      val responseCallback: Consumer[ImapAsyncResponse] = responses => {
+        import collection.JavaConverters._
+
+        val responsesList = ImapResponses(responses.getResponseLines.asScala.to[Seq])
+        logger.trace(s"On response for $userId :\n ${responsesList.mkString("\n")}")
+        self !  Response.SubscribeResponse(responsesList)}
+      val errorCallback: Consumer[Exception] = e => {
+        logger.error("SubscribeHandler command failed", e)
+      }
```
Comment (chibenwa) [inline src/main/scala-2.12/com/linagora/gatling/imap/protocol/command/SubscribeHandler.scala:32]: Please propagate errors to the sender cf https://github.com/linagora/gatling-imap/pull/62/commits/e3765a35c48890ecd8b9f676a646e6b08da72b63

Here and in all your other handlers...

File: src/main/scala-2.12/com/linagora/gatling/imap/action/ImapActionBuilder.scala [hunk 5]
```diff
@@ -186,6 +189,15 @@ case class ImapUnsubscribeActionBuilder(requestName: String, mailbox: Expression
   override val actionName: String = "unsubscribe-action"
 }
 
+case class ImapCreateFolderActionBuilder(requestName: String, mailbox: Expression[String], private val checks: Seq[ImapCheck]) extends ImapCommandActionBuilder {
+  def check(checks: ImapCheck*): ImapCreateFolderActionBuilder = copy(checks = this.checks ++ checks)
+
+  override def props(ctx: ImapActionContext): Props =
+    UnsubscribeAction.props(ctx, requestName, checks, mailbox)
```
Comment (chibenwa) [inline src/main/scala-2.12/com/linagora/gatling/imap/action/ImapActionBuilder.scala]: CreateFolderAction

File: README.adoc [hunk 6]
```diff
@@ -36,6 +36,7 @@ For now only the following commands are supported:
  - DELETE FOLDER
  - RENAME FOLDER
  - EXAMINE FOLDER
+ - MOVE
```
Comment (chibenwa) [inline README.adoc:39]: RFC number please https://datatracker.ietf.org/doc/html/rfc6851

File: src/main/scala-2.12/com/linagora/gatling/imap/protocol/command/GetQuotaHandler.scala [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+      val errorCallback: Consumer[Exception] = _ => {}
+
+      val future = session.execute(new GetQuotaCommand(mailbox))
+      future.setDoneCallback(responseCallback)
+      future.setExceptionCallback(errorCallback)
+  }
+
+  def waitCallback(sender: ActorRef): Receive = {
+    case msg@Response.QuotaResponse(_) =>
+      sender ! msg
```
Comment (chibenwa) [inline src/main/scala-2.12/com/linagora/gatling/imap/protocol/command/GetQuotaHandler.scala:37]: No trace ?

**Review by chibenwa — COMMENTED** [review]
Please don't forget UID MOVE.

File: src/main/scala-2.12/com/linagora/gatling/imap/scenario/ImapGetQuotaScenario.scala (hunk 2 above)
Comment (chibenwa) [reply src/main/scala-2.12/com/linagora/gatling/imap/scenario/ImapGetQuotaScenario.scala:16]: Mhhh acceptable to me. James dissallow setting quota so...

File: src/main/scala-2.12/com/linagora/gatling/imap/protocol/command/CompressCommand.scala [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+  override def getCommandLineBytes: ByteBuf = {
+    val sb = Unpooled.buffer(CompressCommandType.getType.getBytes(StandardCharsets.US_ASCII).length + algorithm.getBytes(StandardCharsets.US_ASCII).length + 100)
+    sb.writeBytes(CompressCommandType.getType.getBytes(StandardCharsets.US_ASCII))
+    sb.writeBytes(" ".getBytes(StandardCharsets.US_ASCII))
+    sb.writeBytes(algorithm.getBytes(StandardCharsets.US_ASCII))
+    sb.writeBytes("\r\n".getBytes(StandardCharsets.US_ASCII))
+    sb
+  }
+
+  override def getCommandType: ImapCommandType = SetQuotaCommandType
```
Comment (Arsnael) [inline src/main/scala-2.12/com/linagora/gatling/imap/protocol/command/CompressCommand.scala:22]:
```suggestion
  override def getCommandType: ImapCommandType = CompressCommandType
```

**Review by Arsnael — APPROVED** [review]

File: src/main/scala-2.12/com/linagora/gatling/imap/protocol/command/CompressCommand.scala [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+import java.nio.charset.StandardCharsets
+
+import com.yahoo.imapnio.async.request.{ImapCommandType, ImapRequestAdapter}
+import io.netty.buffer.{ByteBuf, Unpooled}
+
+case object CompressCommandType extends ImapCommandType {
+  override def getType: String = "COMPRESS"
+}
+
+class CompressCommand(algorithm: String) extends ImapRequestAdapter {
```
Comment (chibenwa) [inline src/main/scala-2.12/com/linagora/gatling/imap/protocol/command/CompressCommand.scala:12]: That's not enough as the client side also need to set up compression.

Actually the ImapNIO session have a dedicated method for compression, that do not rely on commands.

File: src/main/scala-2.12/com/linagora/gatling/imap/protocol/command/StartTLSCommand.scala [hunk 10]
```diff
@@ -0,0 +1,11 @@
+package com.linagora.gatling.imap.protocol.command
+
+import com.yahoo.imapnio.async.request.{AbstractNoArgsCommand, ImapCommandType}
+
+case object StartTLSCommandType extends ImapCommandType {
+  override def getType: String = "STARTTLS"
+}
+
+class StartTLSCommand extends AbstractNoArgsCommand(StartTLSCommandType.getType) {
+  override def getCommandType: ImapCommandType = StartTLSCommandType
```
Comment (chibenwa) [inline src/main/scala-2.12/com/linagora/gatling/imap/protocol/command/StartTLSCommand.scala:10]: Idem there is a dedicated method for starttls: sending the command is not enough you also need to modify the incoming pipeline to decrypt/encrypt SSL.

File: src/main/scala-2.12/com/linagora/gatling/imap/scenario/ImapCompressScenario.scala [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+import io.gatling.core.feeder.FeederBuilder
+import io.gatling.core.structure.ScenarioBuilder
+
+object ImapCompressScenario {
+  def apply(feeder: FeederBuilder): ScenarioBuilder =
+    scenario("Imap")
+      .feed(feeder)
+      .exec(imap("Connect").connect()).exitHereIfFailed
+      .exec(imap("capability").capability().check(contains("COMPRESS=DEFLATE")))
+      .exec(imap("compress").compress("deflate").check(ok))
```
Comment (chibenwa) [inline src/main/scala-2.12/com/linagora/gatling/imap/scenario/ImapCompressScenario.scala:14]: Please add a noop command after to ensure the connection is still usable once compression is turned on.

File: src/main/scala-2.12/com/linagora/gatling/imap/scenario/ImapStartTLSScenario.scala [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+import io.gatling.core.feeder.FeederBuilder
+import io.gatling.core.structure.ScenarioBuilder
+
+object ImapStartTLSScenario {
+  def apply(feeder: FeederBuilder): ScenarioBuilder =
+    scenario("Imap")
+      .feed(feeder)
+      .exec(imap("Connect").connect()).exitHereIfFailed
+      .exec(imap("capability").capability().check(contains("STARTTLS")))
+      .exec(imap("startTLS").startTLS().check(ok))
```
Comment (chibenwa) [inline src/main/scala-2.12/com/linagora/gatling/imap/scenario/ImapStartTLSScenario.scala:14]: Please add a noop command after to ensure the connection is still usable once encryption is turned on.

File: src/main/scala-2.12/com/linagora/gatling/imap/protocol/command/StartTLSCommand.scala (hunk 10 above)
Comment (quantranhong1999) [reply src/main/scala-2.12/com/linagora/gatling/imap/protocol/command/StartTLSCommand.scala:10]: I checked and there is no dedicated starttls method in imapnio. I also give a quick try modifying the SSL context in `ImapSession` but did not succeed. I will give it a try later.

Comment (chibenwa) [reply src/main/scala-2.12/com/linagora/gatling/imap/protocol/command/StartTLSCommand.scala:10]: You are right, it seems not implemented.

I'm not sure StartTLS is very valuable:

 - Not part of our target deployments (that uses implicit TLS)
 - And hard to implement as it is not supported by IMAP nio
 
 I would forget about it for now.

**Pull request merged**